src/libos/crates/async-rt/src/scheduler/local_scheduler/mod.rs

```diff
@@ -0,0 +1,364 @@
+
+                // Notify the idle status as early as possible
+                debug_assert!(old_len >= 1);
+                let is_idle = old_len == 1;
+                if is_idle {
+                    self.status_notifier
+                        .notify_idle_status(self.this_vcpu, true);
+                }
+
+                return Some(entity);
+            }
+
+            self.reschedule(&mut rqs);
+        }
+    }
+
+    fn reschedule(&self, rqs: &mut RqsLockGuard<'_, E>) {
+        // All entities in the back runqueues have assigned timeslices
+        // when they are enqueued. So now we only need to switch
+        // the front and back runqueues.
+        rqs.switch();
+    }
+
+    /// Drain entities that satisfy a condition.
+    ///
+    /// Similar to the `dequeue` method, the `drain` method removes entities
+    /// from the scheduler.
+    ///
+    /// This method is useful for migration tasks from one vCPU to another
+    /// during load balancing.
+    pub fn drain<F>(&self, mut cond: F, drained: &mut Vec<Arc<E>>) -> usize
+    where
+        F: FnMut(&E) -> bool,
+    {
+        if self.len.load(Relaxed) == 0 {
+            debug!("src scheduler length is zero");
+            return 0;
+        }
+
+        let max_drained = drained.capacity() - drained.len();
+        debug!("max_drained is {:?}", &max_drained);
+        debug!(
+            "the current scheduler length is {:?}",
+            self.len.load(Relaxed)
+        );
+        let mut num_drained = 0;
+        let mut rqs = self.lock_rqs();
+        let mut drain_queue = |rqs: &mut RunQueues<E>| {
+            if num_drained >= max_drained {
+                return;
+            }
+
+            // Iterate the internal run queues in order of ascending priority
+            for (q_i, q) in rqs.run_queues.iter_mut().rev().enumerate() {
+                let mut i = 0;
+                debug!("the queue index and length is {:?}", (31 - q_i, q.len()));
+                while i < q.len() {
+                    let entity = &q[i];
+                    if cond(entity) {
+                        let entity = q.swap_remove_back(i).unwrap();
+                        entity.sched_state().clear_enqueued();
+                        self.len.fetch_sub(1, Relaxed);
+
+                        if q.is_empty() {
+                            let idx = 31 - q_i;
+                            rqs.nonempty_mask &= !(1 << idx);
+                        }
+                        drained.push(entity);
+                        num_drained += 1;
+                        if num_drained >= max_drained {
+                            return;
+                        }
+                    } else {
+                        i += 1;
+                    }
+                }
+            }
+        };
+        drain_queue(rqs.back_mut());
+        drain_queue(rqs.front_mut());
+        num_drained
+    }
+
+    pub fn wait_enqueue(&self) {
+        let mut count = 5_000_000;
```

Comment:
Because the vcpu_selector is inclined to choose **active** vcpus (not sleeped) to enqueue tasks. In other words, those sleeped vcpus would never be unparked. For instance, the initial process are going to spawn some computing-intensive tasks, but those vcpus has slumped into sleeping state when start enqueueing tasks, the large amout of computing resource would be wasted.
Therefore, the vcpus requires more time for idle states to avoid slumping into sleep states too quickly.

Reply:
You see. It is non-trivial to see why choosing a very big number here. Your response should be added to the code as comments.

Your argument makes sense. But I think it reveals a bigger problem. Is the vCPU selector's current policy suboptimal? Maybe the vCPU selector should prefer sleeping vCPUs over active vCPUs. Assigning a task to sleeping vCPUs incurs larger latencies, but it activates more vCPUs, which is good for the overall performance of the system.